#### Open and read the file `Hawking-Questioning-the-Universe.txt`.

In [60]:
f = open("Hawking-Questioning-the-Universe.txt", "r")
print(f.read())
f.close()

There is nothing bigger or older than the universe. The questions I would like to talk about are: one, where did we come from? How did the universe come into being? Are we alone in the universe? Is there alien life out there? What is the future of the human race? Up until the 1920s, everyone thought the universe was essentially static and unchanging in time. Then it was discovered that the universe was expanding. Distant galaxies were moving away from us. This meant they must have been closer together in the past. If we extrapolate back, we find we must have all been on top of each other about 15 billion years ago. This was the Big Bang, the beginning of the universe. But was there anything before the Big Bang? If not, what created the universe? Why did the universe emerge from the Big Bang the way it did? We used to think that the theory of the universe could be divided into two parts. First, there were the laws like Maxwell’s equations and general relativity that determined the evolu

#### Print the first 50 characters.

In [9]:
f = open("Hawking-Questioning-the-Universe.txt", "r")
print(f.read(50))
f.close()

There is nothing bigger or older than the universe


#### Print from the 52nd to the 163rd character.

In [11]:
f = open("Hawking-Questioning-the-Universe.txt", "r")
f.read(52)
print(f.read(111))
f.close()

The questions I would like to talk about are: one, where did we come from? How did the universe come into being


---

#### Open the file `Hawking-Questioning-the-Universe.txt` and create a list named `text_as_list` where every word is an element of the list. 

Hint: there are two ways to do it. Either with `.readlines()` or with `.split()`. I encourage you to google how both work, but I suggest to adopt the latter.

Hint: if you want to split the text so every word is an element of a list, what should be the parameter of the `split()` method? Read more [here](https://www.w3schools.com/python/ref_string_split.asp).

In [2]:
f = open("Hawking-Questioning-the-Universe.txt", "r")
text = f.read()
text_as_list = text.split(" ")
print(text_as_list)

['There', 'is', 'nothing', 'bigger', 'or', 'older', 'than', 'the', 'universe.', 'The', 'questions', 'I', 'would', 'like', 'to', 'talk', 'about', 'are:', 'one,', 'where', 'did', 'we', 'come', 'from?', 'How', 'did', 'the', 'universe', 'come', 'into', 'being?', 'Are', 'we', 'alone', 'in', 'the', 'universe?', 'Is', 'there', 'alien', 'life', 'out', 'there?', 'What', 'is', 'the', 'future', 'of', 'the', 'human', 'race?', 'Up', 'until', 'the', '1920s,', 'everyone', 'thought', 'the', 'universe', 'was', 'essentially', 'static', 'and', 'unchanging', 'in', 'time.', 'Then', 'it', 'was', 'discovered', 'that', 'the', 'universe', 'was', 'expanding.', 'Distant', 'galaxies', 'were', 'moving', 'away', 'from', 'us.', 'This', 'meant', 'they', 'must', 'have', 'been', 'closer', 'together', 'in', 'the', 'past.', 'If', 'we', 'extrapolate', 'back,', 'we', 'find', 'we', 'must', 'have', 'all', 'been', 'on', 'top', 'of', 'each', 'other', 'about', '15', 'billion', 'years', 'ago.', 'This', 'was', 'the', 'Big', 'Bang

#### This code will remove the punctuation from your list and applies the lower() method that returns a string where all characters are lower case. Simple run the cell.

In [9]:
import string
text_as_list = [i.translate(str.maketrans('', '', string.punctuation)).lower() for i in text_as_list]

#### Create a new list `words_begin_with_a` and append all words of the `text_as_list` that begin with a small _a_.

Hint: you have to define first an empty list.

Hint: in order to find if a string starts with one or more letters, you should use the method `.startswith()`. Read more [here](https://www.w3schools.com/python/ref_string_startswith.asp).

In [10]:
words_begin_with_a = []

for words in text_as_list:
    if words.startswith("a"):
        words_begin_with_a.append(words)

print(words_begin_with_a)


['about', 'are', 'are', 'alone', 'alien', 'and', 'away', 'all', 'about', 'ago', 'anything', 'and', 'all', 'at', 'and', 'and', 'all', 'about', 'and', 'and', 'and', 'and', 'allow', 'another', 'and', 'and', 'also', 'a', 'are', 'agreement', 'an', 'and', 'are', 'alone', 'arose', 'appear', 'a', 'appeared', 'appearing', 'algae', 'ago', 'ago', 'and', 'about', 'appeared', 'a', 'a', 'appearing', 'available', 'aliens', 'am', 'appear', 'and', 'a', 'and', 'aliens', 'a', 'an', 'any', 'alien', 'are', 'alien', 'at', 'a', 'a', 'an', 'against', 'abduction', 'aliens', 'a', 'are', 'and', 'are', 'an', 'and', 'are', 'along', 'ability', 'and', 'aggressive', 'advantage', 'avoid', 'alone', 'answers', 'am', 'all', 'and', 'answers', 'a', 'a', 'and', 'are']


Answer: 310

#### Create a counter and use a while loop to counter how many words starts with a _u_.

In [11]:
i = 0
counter = 0
other_counter = 0

while i<len(text_as_list):
    if text_as_list[i].startswith('u'):
        counter+=1
    else:
        other_counter+=1
    i += 1
print(counter)

30


Answer: 30

#### Create a counter and use a while loop to counter how many words do NOT start with a _u_.

In [12]:
print(other_counter)

877


Answer: 877

#### Write an if statement that checkes if the word "universe" appears in the text. Feel free to google how we test if a word is _in_ a Python list.

In [41]:
a = text_as_list.count("universe")
if a>0:
    True
    print("True")
else:
    False
    print("False")

True


#### Write an if statement that checkes if the word "alliens" appears in the text. If not print "the word does not appear"

In [44]:
a = text_as_list.count("alliens")
b = text_as_list.count("the word does not appear")
if a==0:
    print("The word does not appear")
    if b==0:
        print("I didn't find the words")
    

The word does not appear
I didn't find the words


#### Write an if statement that checkes if the word "alliens" appears in the text. If not, check if the word "International Space Station" appears in the text. If neither is in the text, print "I didn't find the words".

#### Take the text_as_list and create a dictionary named `mydictionary`. The words of the text should be the keys of the dictionary and the index of the word the values of the keys.

Hint: there is a method in Python that adds a counter to an iterable.

In [13]:
dic = {}
total = len(text_as_list)
a = 1

for i in text_as_list:
    dic[i] = a
    a+=1
    
    
print(dic)




{'there': 624, 'is': 891, 'nothing': 314, 'bigger': 4, 'or': 832, 'older': 6, 'than': 879, 'the': 896, 'universe': 897, 'questions': 856, 'i': 857, 'would': 564, 'like': 281, 'to': 882, 'talk': 16, 'about': 479, 'are': 900, 'one': 529, 'where': 20, 'did': 151, 'we': 899, 'come': 29, 'from': 460, 'how': 434, 'into': 786, 'being': 497, 'alone': 762, 'in': 827, 'alien': 627, 'life': 843, 'out': 785, 'what': 135, 'future': 818, 'of': 895, 'human': 670, 'race': 671, 'up': 52, 'until': 236, '1920s': 55, 'everyone': 372, 'thought': 57, 'was': 526, 'essentially': 61, 'static': 62, 'and': 898, 'unchanging': 64, 'time': 878, 'then': 67, 'it': 872, 'discovered': 70, 'that': 862, 'expanding': 75, 'distant': 76, 'galaxies': 77, 'were': 741, 'moving': 79, 'away': 80, 'us': 82, 'this': 656, 'meant': 84, 'they': 565, 'must': 407, 'have': 858, 'been': 867, 'closer': 89, 'together': 90, 'past': 747, 'if': 807, 'extrapolate': 96, 'back': 97, 'find': 852, 'all': 840, 'on': 779, 'top': 106, 'each': 108, 'o

In [ ]:
mydictionary = {}

for i,j in enumerate(text_as_list):
    mydictionary[j] = i
    
print(len(mydictionary))

#### Create the same dictonary but now index of  words in the text should be the dictionary's keys and the words should be the key's values.

In [14]:
dic = {}
total = len(text_as_list)
a = 1

for i in text_as_list:
    dic[a] = i
    a+=1
    
    
print(dic)

{1: 'there', 2: 'is', 3: 'nothing', 4: 'bigger', 5: 'or', 6: 'older', 7: 'than', 8: 'the', 9: 'universe', 10: 'the', 11: 'questions', 12: 'i', 13: 'would', 14: 'like', 15: 'to', 16: 'talk', 17: 'about', 18: 'are', 19: 'one', 20: 'where', 21: 'did', 22: 'we', 23: 'come', 24: 'from', 25: 'how', 26: 'did', 27: 'the', 28: 'universe', 29: 'come', 30: 'into', 31: 'being', 32: 'are', 33: 'we', 34: 'alone', 35: 'in', 36: 'the', 37: 'universe', 38: 'is', 39: 'there', 40: 'alien', 41: 'life', 42: 'out', 43: 'there', 44: 'what', 45: 'is', 46: 'the', 47: 'future', 48: 'of', 49: 'the', 50: 'human', 51: 'race', 52: 'up', 53: 'until', 54: 'the', 55: '1920s', 56: 'everyone', 57: 'thought', 58: 'the', 59: 'universe', 60: 'was', 61: 'essentially', 62: 'static', 63: 'and', 64: 'unchanging', 65: 'in', 66: 'time', 67: 'then', 68: 'it', 69: 'was', 70: 'discovered', 71: 'that', 72: 'the', 73: 'universe', 74: 'was', 75: 'expanding', 76: 'distant', 77: 'galaxies', 78: 'were', 79: 'moving', 80: 'away', 81: 'fro

In [1]:
mydictionary = {}

for i,j in enumerate(text_as_list):
    mydictionary[i] = j
    
print(len(mydictionary))

NameError: name 'text_as_list' is not defined

#### Why the lenghts of the two dictionaries are not the same?

#### Create a dictionary `dict_counter` that takes as keys the words of text_as_list. The values of each key should be the number that each words is attested in the list.

Hint: Check the [`count()` method](https://www.w3schools.com/python/ref_list_count.asp).

In [5]:
dic = {}
total = len(text_as_list)

for i in text_as_list:
    dic[i] = text_as_list.count(i)
    
    
print(dic)

{'There': 1, 'is': 12, 'nothing': 1, 'bigger': 1, 'or': 5, 'older': 1, 'than': 2, 'the': 71, 'universe.': 5, 'The': 6, 'questions': 2, 'I': 7, 'would': 3, 'like': 3, 'to': 22, 'talk': 1, 'about': 4, 'are:': 1, 'one,': 1, 'where': 1, 'did': 3, 'we': 21, 'come': 2, 'from?': 1, 'How': 1, 'universe': 9, 'into': 4, 'being?': 1, 'Are': 1, 'alone': 2, 'in': 14, 'universe?': 3, 'Is': 1, 'there': 8, 'alien': 3, 'life': 9, 'out': 3, 'there?': 1, 'What': 1, 'future': 3, 'of': 39, 'human': 2, 'race?': 1, 'Up': 1, 'until': 2, '1920s,': 1, 'everyone': 2, 'thought': 1, 'was': 10, 'essentially': 1, 'static': 1, 'and': 18, 'unchanging': 1, 'time.': 2, 'Then': 1, 'it': 8, 'discovered': 1, 'that': 11, 'expanding.': 1, 'Distant': 1, 'galaxies': 1, 'were': 3, 'moving': 1, 'away': 1, 'from': 3, 'us.': 1, 'This': 6, 'meant': 1, 'they': 2, 'must': 3, 'have': 14, 'been': 6, 'closer': 1, 'together': 1, 'past.': 2, 'If': 5, 'extrapolate': 1, 'back,': 1, 'find': 2, 'all': 3, 'on': 8, 'top': 1, 'each': 1, 'other':

#### Without printing the whole dictionary, find how many times the word "question" appears in the list by using your new dictiornary

In [23]:
print(dic["question"])

2


#### Create a pandas dataframe from the items of the dictionary

In [27]:
import pandas as pd
df = pd.DataFrame(dic.items())
  
df

0   1
0         There   1
1            is  12
2       nothing   1
3        bigger   1
4            or   5
..          ...  ..
411      making   1
412   progress.   1
413       Thank   1
414         you   1
415  listening.   1

[416 rows x 2 columns]

#### Get the first row of your dataframe

In [28]:
df.iloc[0]
df.head(1)

0  1
0  There  1

#### Get the first column of your dataframe

In [29]:
df[1]

0       1
1      12
2       1
3       1
4       5
       ..
411     1
412     1
413     1
414     1
415     1
Name: 1, Length: 416, dtype: int64

#### By using the `pandas.DataFrame.loc`, find how many times the words galaxies and galaxies appear in the dataframe

In [32]:
df.reset_index(inplace=True)
df.set_index(1, inplace=True)
df.loc['galaxy']
df.loc['galaxies']



KeyError: 'galaxy'

#### Write a recursion in a function that prints 5! (i.e., the factorial of 5).

In [26]:
df

0
1             
1        There
12          is
1      nothing
1       bigger
5           or
..         ...
1       making
1    progress.
1        Thank
1          you
1   listening.

[416 rows x 1 columns]

In [44]:
def fact(n):
    if n==0:
        return 1
    else:
        return n * fact(n-1)
    

In [43]:
print(fact(5))

120


#### Write a recursion in a function that prints the $n^{th}$ [Fibonacci number](https://en.wikipedia.org/wiki/Fibonacci_number):

1, 1, 2, 3, 5, 8, 13, 21, 34

In [55]:
def fib(n):
    if (n<=1):
        return n
    return fib (n-2) + fib(n-1)

#### Which is the $20^{th}$ Fibonacci number?

In [58]:
fib(8)

21

_We will try together tail recursion._